### PR TITLE
Issue #13: Provide plugins for providers.

### LINKED ISSUE #13: GitLab support
It's quite a nice initiative and it'd be great to see it supporting GitLab CI

Some code has already been written for that purpose
https://github.com/shulard/laravel-build-artifacts
but having a composer plugin solves the problem of `composer` not executing dependencies' scripts.

Anyway support for GitLab needs something more than simply defining an URL in `composer.json`
A such, support object inside the `dist` key would be a first step to support forward-compatibility.

https://github.com/shulard/laravel-build-artifacts/pull/1 makes (optional) use of `ref` and `tag`, `stage`, `(job-)name` and the `private token` (which can be passed through environment).

Do you think there is a place for this inside `composer-artifacts`?

### PR DESCRIPTION
**Pull request by drupol**
Fix #13

### CODE REVIEW

**Comment by drupol** [comment]
Thanks for your feedback, I've updated the PR based on it.
Don't hesitate to share the Gitlab plugin you made, it might probably end up in this repo as well :)

**Comment by drzraf** [comment]
I sure will ^

Since my personal use-case is about downloading a tarball of assets rather than the full project I'd like to (also) support `POST_PACKAGE_*` `PackageEvent` (#14), but I not yet sure how to do that without creating complexity.

Another API modification is that in order to use `$io`, not only `$plugin` needed to be protected rather than private, but `$io` needs to be public.

```diff
diff --git a/src/Plugin.php b/src/Plugin.php
index b28a52f..4ddf4c0 100644
--- a/src/Plugin.php
+++ b/src/Plugin.php
@@ -31,7 +31,7 @@ class Plugin implements PluginInterface, EventSubscriberInterface
      *
      * @var IOInterface
      */
-    private $io;
+    public $io;
 
     /**
      * Get the configuration.
```

**Comment by drupol** [comment]
I just added the public method `::getIo()`, would it be enough ?

File: src/Plugin.php [hunk 1]
```diff
@@ -43,134 +65,116 @@ public function getConfig()
...
 
-        /** @var Package $package */
-        $package = $operation->getTargetPackage();
+        if (!isset($config[$package->getName()])) {
+            return;
+        }
 
-        if (\array_key_exists($package->getName(), $this->getConfig())) {
-            $this->updatePackageConfiguration($package);
+        $packageCconfig = $config[$package->getName()];
```
Comment (drzraf) [inline src/Plugin.php]: `packageCconfig` -> `packageConfig`

Comment (drupol) [reply src/Plugin.php]: Fixed.

File: src/Plugin.php [hunk 2]
```diff
@@ -32,145 +32,120 @@ class Plugin implements PluginInterface, EventSubscriberInterface
...
+        /** @var \Composer\DependencyResolver\Operation\OperationInterface $operation */
         $operation = $event->getOperation();
 
-        /** @var Package $package */
-        $package = $operation->getPackage();
-
-        if (\array_key_exists($package->getName(), $this->getConfig())) {
-            $this->updatePackageConfiguration($package);
+        switch ($event->getName()) {
+            case 'post-package-update':
```
[the remark this answers is not in the document]
Comment (drupol) [reply src/Plugin.php]: As this setting is now configurable is the YAML, I would use `(pre|post)-package-(install|update)`, ok for you ?

**Comment by drupol** [comment]
`$provider->getIo()` could be replaced with `$provider->getPlugin()->getIo()`

**Comment by drzraf** [comment]
anything blocking?

**Comment by drupol** [comment]
> anything blocking?

Hi Rafael,

I will create a new project based on what we've done here, including by default the provider plugins.

In the meantime, I will closed this issue.

**Pull request closed without merging**